src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -412,6 +412,19 @@ internal static class ConstraintsHelper
                     containingSymbol.CheckConstraintTypeVisibility(syntax.Location, constraintType, diagnostics);
                     constraintTypeBuilder.Add(constraintType);
                 }
+
+                if (constraintType.ContainsNullableReferenceTypes())
+                {
+                    // Note: Local functions register their need for the Nullable attribute on constraints in LocalRewriter
+                    bool onLocalFunction = containingSymbol.Kind == SymbolKind.Method && ((MethodSymbol)containingSymbol).MethodKind == MethodKind.LocalFunction;
+                    containingSymbol.DeclaringCompilation.EnsureNullableAttributeExists(diagnostics, syntax.Location, modifyCompilation: !onLocalFunction);
```

Comment:
It seems strange that `CheckConstraints` now has the side-effect of ensuring the `NullableAttribute` is emitted. Perhaps add a `// PROTOTYPE` comment to consider moving this out of `CheckConstraints`.

Reply:
Every other place where we use `ReportMissingNonNullTypesContextForAnnotation`, we also do `EnsureNullableAttributeExists` along with it.
Do you think we should move both? Where we would move to?

---
In reply to: [209111841](https://github.com/dotnet/roslyn/pull/29181#discussion_r209111841) [](ancestors = 209111841)